storage.go

```diff
@@ -3,34 +3,187 @@ package ftp
 import (
 	"context"
 	"io"
+	"path/filepath"
 
+	"github.com/beyondstorage/go-storage/v4/pkg/iowrap"
 	. "github.com/beyondstorage/go-storage/v4/types"
+	"github.com/jlaffaye/ftp"
+	mime "github.com/qingstor/go-mime"
 )
 
+type listDirInput struct {
+	rp  string
+	dir string
+
+	started           bool
+	continuationToken string
+	objChan           []*ftp.Entry
+	counter           int
+	buf               []byte
+}
+
 func (s *Storage) create(path string, opt pairStorageCreate) (o *Object) {
-	panic("not implemented")
+	if opt.HasObjectMode && opt.ObjectMode.IsDir() {
+		o = s.newObject(false)
+		o.Mode = ModeDir
+	} else {
+		o = s.newObject(false)
+		o.Mode = ModeRead
+	}
+
+	o.ID = filepath.Join(s.workDir, path)
+	o.Path = path
+	return o
+}
+func (s *Storage) createDir(ctx context.Context, path string) (o *Object, err error) {
+	rp := s.getAbsPath(path)
+	s.connection.ChangeDir(s.workDir)
+	err = s.connection.MakeDir(rp)
+	if err != nil {
+		return
+	}
+
+	o = s.newObject(true)
+	o.ID = rp
+	o.Path = path
+	o.Mode |= ModeDir
+	return
 }
 
 func (s *Storage) delete(ctx context.Context, path string, opt pairStorageDelete) (err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+	s.connection.ChangeDir(filepath.Dir(rp))
+	err = s.connection.Delete(filepath.Base(rp))
+	s.connection.ChangeDir(s.workDir)
+	if err != nil {
+		return err
+	}
+	return
+}
+func (input *listDirInput) ContinuationToken() string {
+	return input.continuationToken
 }
 
 func (s *Storage) list(ctx context.Context, path string, opt pairStorageList) (oi *ObjectIterator, err error) {
-	panic("not implemented")
+	//rp := s.getAbsPath(path)
+	input := listDirInput{
+		// Always keep service original name as rp.
+		rp: s.getAbsPath(path),
+		// Then convert the dir to slash separator.
+		dir: filepath.ToSlash(path),
+
+		// if HasContinuationToken, we should start after we scanned this token.
+		// else, we can start directly.
+		started:           !opt.HasContinuationToken,
+		continuationToken: opt.ContinuationToken,
+
+		buf: make([]byte, 8192),
+	}
+
+	return NewObjectIterator(ctx, s.listDirNext, &input), nil
 }
 
 func (s *Storage) metadata(opt pairStorageMetadata) (meta *StorageMeta) {
-	panic("not implemented")
+	meta = NewStorageMeta()
+	meta.WorkDir = s.workDir
+	return meta
+
 }
 
 func (s *Storage) read(ctx context.Context, path string, w io.Writer, opt pairStorageRead) (n int64, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+
+	if err != nil {
+		return 0, err
+	}
+
+	r, err := s.connection.Retr(rp)
+	if opt.HasOffset {
```

Comment:
Maybe it's better to use `if-else` here to avoid redundant api call? If that, we should declare the variable `r` before the `if-else` block.

Reply:
Or, since the `Retr` is implemented by `RetrFrom`, you can also set offset's default as 0, and check `opt.HasOffset` to set offset, then just call `RetrFrom`.